http_service/http_handler.go

```diff
@@ -0,0 +1,136 @@
+	"github.com/gin-gonic/gin"
+	"github.com/gin-gonic/gin/binding"
+	"net/http"
+)
+
+type HttpHandler struct {
+	bookService     book_Service.BookService
+	activityHandler cache.ActivityCacher
+}
+
+func NewHttpHandler(bookService book_Service.BookService, activityHandler cache.ActivityCacher) HttpHandler {
+	return HttpHandler{
+		bookService:     bookService,
+		activityHandler: activityHandler,
+	}
+}
+
+type username struct {
+	Username string `json:"username"`
+}
+
+func (h *HttpHandler) AddBook(c *gin.Context) {
+	var book models.Book
+	if err := c.ShouldBindBodyWith(&book, binding.JSON); err != nil {
+		c.JSON(http.StatusBadRequest, "error: couldn't bind body in AddBook request")
+		return
+	}
+
+	id, err := h.bookService.InsertBook(context.Background(), &book)
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	c.JSON(http.StatusOK, gin.H{"id": id})
+}
+
+func (h *HttpHandler) UpdateBook(c *gin.Context) {
+	var book models.Book
+	err := c.ShouldBindBodyWith(&book, binding.JSON)
+	if err != nil {
+		c.JSON(http.StatusBadRequest, "error: couldn't bind body in UpdateBook request")
+		return
+	}
+
+	err = h.bookService.UpdateBook(context.Background(), book.Title, book.Id)
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	c.JSON(http.StatusBadRequest, "book updated")
+}
+
+func (h *HttpHandler) GetBook(c *gin.Context) {
+	s, err := h.bookService.GetBook(context.Background(), c.Query(elastic_fields.Id))
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	c.JSON(http.StatusOK, s)
+}
+
+func (h *HttpHandler) DeleteBook(c *gin.Context) {
+	err := h.bookService.DeleteBook(context.Background(), c.Query(elastic_fields.Id))
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	c.JSON(http.StatusBadRequest, "book deleted")
+}
+
+func (h *HttpHandler) Search(c *gin.Context) {
+	searchParams := make(map[string]string)
+	searchParams["title"] = c.Query(elastic_fields.Title)
+	searchParams["author"] = c.Query(elastic_fields.Author)
+	searchParams["price_range"] = c.Query(query_fields.PriceRange)
+	s, err := h.bookService.Search(context.Background(), searchParams)
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	c.JSON(http.StatusOK, s)
+}
+
+func (h *HttpHandler) StoreInfo(c *gin.Context) {
+	info, err := h.bookService.StoreInfo(context.Background())
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	c.JSON(http.StatusOK, gin.H{
+		"store_info": info,
+	})
+}
+
+func (h *HttpHandler) Activity(c *gin.Context) {
+
+	actions, err := h.activityHandler.GetLastActions(c.Query(query_fields.UserName))
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+	}
+
+	c.JSON(http.StatusOK, actions)
+}
+
+func (h *HttpHandler) AddAction(c *gin.Context) {
```

Comment:
Interesting solution. can you try and implement it with gin middleware ([exp](https://sosedoff.com/2014/12/21/gin-middleware.html))?

Reply:
I used middleware in this solution, the AddAction is actually a middleware function which I enter to my router with the gin use function. I added a line c.next() int the end, but I'm not sure if it's necessary, since it worked well before this addition.

Reply:
Please check go-searcher middleware, and how we init it..

Reply:
done :)